File: cursor-plugins/dev-team/utils/archi-decide.py
#!/usr/bin/env python3
"""
archi-decide.py — Architecture Decision Record (ADR) management utility.

Commands:
  create    Create a new ADR
  list      List/query ADRs
  update    Update ADR status

Usage:
  python archi-decide.py create --title "..." --background "..." --decision "..." ...
  python archi-decide.py list [--status accepted]
  python archi-decide.py update --file "..." --status "accepted"
"""

import argparse
import json
import os
import re
import sys
from datetime import datetime


DECISIONS_DIR = os.path.join("openspec", "architecture", "decisions")
TEMPLATE_PATH = os.path.join("plugins", "dev-team", "templates", "adr.md")

VALID_STATUSES = ("proposed", "accepted", "deprecated", "superseded")


def resolve_project_root():
    """Find project root by locating openspec/ or .git."""
    cwd = os.getcwd()
    for start in [cwd] + list(_walk_up(cwd)):
        if os.path.isdir(os.path.join(start, "openspec")):
            return start
        if os.path.isdir(os.path.join(start, ".git")):
            return start
    return cwd


def _walk_up(path):
    while True:
        parent = os.path.dirname(path)
        if parent == path:
            break
        yield parent
        path = parent


def _slugify(title):
    """Convert a title to kebab-case for filename use."""
    slug = title.lower().strip()
    slug = re.sub(r'[^\w\s-]', '', slug)
    slug = re.sub(r'[\s_]+', '-', slug)
    slug = re.sub(r'-+', '-', slug)
    return slug.strip('-')


def create_adr(project_root, title, background, decision, consequences, alternatives, scope, status="proposed"):
    """Create a new ADR file.

    Args:
        project_root: Project root directory
        title: ADR title
        background: Background context
        decision: The decision
        consequences: Consequences text
        alternatives: List of alternative dicts [{"name":..., "pros":..., "cons":...}]
        scope: Comma-separated model element IDs
        status: Initial status (default: proposed)

    Returns:
        dict with result info
    """
    if status not in VALID_STATUSES:
        return {"success": False, "error": f"Invalid status '{status}'. Must be one of: {', '.join(VALID_STATUSES)}"}

    date_str = datetime.now().strftime("%Y-%m-%d")
    slug = _slugify(title)
    filename = f"{date_str}-{slug}.md"

    decisions_dir = os.path.join(project_root, DECISIONS_DIR)
    os.makedirs(decisions_dir, exist_ok=True)

    filepath = os.path.join(decisions_dir, filename)

    if os.path.exists(filepath):
        return {"success": False, "error": f"ADR already exists: {filename}"}

    # Build alternatives section
    alt_text = ""
    if isinstance(alternatives, str):
        try:
            alternatives = json.loads(alternatives)
        except json.JSONDecodeError:
            alternatives = []

    if alternatives:
        for i, alt in enumerate(alternatives, 1):
            alt_text += f"\n### 方案 {i}：{alt.get('name', 'Unnamed')}\n"
            alt_text += f"- **描述**: {alt.get('description', alt.get('name', ''))}\n"
            pros = alt.get('pros', [])
            if isinstance(pros, list):
                for p in pros:
                    alt_text += f"- **优点**: {p}\n"
            else:
                alt_text += f"- **优点**: {pros}\n"
            cons = alt.get('cons', [])
            if isinstance(cons, list):
                for c in cons:
                    alt_text += f"- **缺点**: {c}\n"
            else:
                alt_text += f"- **缺点**: {cons}\n"

    # Build scope list
    scope_items = [s.strip() for s in scope.split(",") if s.strip()]
    scope_text = "\n".join(f"- {s}" for s in scope_items) if scope_items else "- (none)"

    content = f"""# ADR: {title}

- **日期**: {date_str}
- **状态**: {status}

## 背景

{background}

## 决策

{decision}

## 后果

### 正面后果
- (见下方描述)

### 负面后果
- (见下方描述)

{consequences}

## 备选方案
{alt_text if alt_text else "\n(无备选方案记录)\n"}

## 影响范围

{scope_text}
"""

    with open(filepath, "w", encoding="utf-8") as f:
        f.write(content)

    return {
        "success": True,
        "path": filepath,
        "filename": filename,
    }


def list_adrs(project_root, status_filter=None):
    """List ADRs, optionally filtered by status.

    Args:
        project_root: Project root directory
        status_filter: Optional status to filter by

    Returns:
        dict with adrs list
    """
    decisions_dir = os.path.join(project_root, DECISIONS_DIR)

    if not os.path.isdir(decisions_dir):
        return {"adrs": [], "count": 0}

    adrs = []
    try:
        for filename in sorted(os.listdir(decisions_dir), reverse=True):
            if not filename.endswith(".md"):
                continue

            filepath = os.path.join(decisions_dir, filename)
            try:
                with open(filepath, "r", encoding="utf-8") as f:
                    content = f.read()
            except OSError:
                continue

            # Extract title
            title_match = re.search(r'^# ADR:\s*(.+)$', content, re.MULTILINE)
            title = title_match.group(1).strip() if title_match else filename

            # Extract status
            status_match = re.search(r'\*\*状态\*\*:\s*(\w+)', content)
            status = status_match.group(1).strip() if status_match else "unknown"

            # Extract date
            date_match = re.search(r'\*\*日期\*\*:\s*(\S+)', content)
            date = date_match.group(1).strip() if date_match else ""

            # Extract scope
            scope_section = re.search(r'## 影响范围\s*\n(.*?)(?=\n##|\Z)', content, re.DOTALL)
            scope = []
            if scope_section:
                scope = [s.strip("- ").strip() for s in scope_section.group(1).strip().split("\n") if s.strip().startswith("-")]

            if status_filter and status != status_filter:
                continue

            adrs.append({
                "filename": filename,
                "title": title,
                "status": status,
                "date": date,
                "scope": scope,
                "path": filepath,
            })

    except OSError:
        pass

    return {"adrs": adrs, "count": len(adrs)}


def update_adr(project_root, filename, status, superseded_by=None):
    """Update an ADR's status.

    Args:
        project_root: Project root directory
        filename: ADR filename (e.g., "2026-05-12-use-postgresql.md")
        status: New status value
        superseded_by: Reference to superseding ADR (required when status=superseded)

    Returns:
        dict with result info
    """
    if status not in VALID_STATUSES:
        return {"success": False, "error": f"Invalid status '{status}'. Must be one of: {', '.join(VALID_STATUSES)}"}

    if status == "superseded" and not superseded_by:
        return {"success": False, "error": "Status 'superseded' requires --superseded-by reference"}

    filepath = os.path.join(project_root, DECISIONS_DIR, filename)

    if not os.path.isfile(filepath):
        return {"success": False, "error": f"ADR not found: {filename}"}

    try:
        with open(filepath, "r", encoding="utf-8") as f:
            content = f.read()
    except OSError as e:
        return {"success": False, "error": f"Failed to read ADR: {e}"}

    # Update status line
    old_status_match = re.search(r'(\*\*状态\*\*:\s*)\w+', content)
    if old_status_match:
        content = content[:old_status_match.start(1)] + f'**状态**: {status}' + content[old_status_match.end():]

        # Add superseded reference if applicable
        if status == "superseded" and superseded_by:
            ref_line = f"\n- **取代者**: {superseded_by}"
            if "**取代者**" not in content:
                # Insert after status line
                status_end = content.find('\n', content.find(f'**状态**: {status}'))
                if status_end > 0:
                    content = content[:status_end] + ref_line + content[status_end:]

    with open(filepath, "w", encoding="utf-8") as f:
        f.write(content)

    return {
        "success": True,
        "path": filepath,
        "old_status": old_status_match.group(1).strip() if old_status_match else "unknown",
        "new_status": status,
    }


def main():
    parser = argparse.ArgumentParser(description="Architecture Decision Record management")
    subparsers = parser.add_subparsers(dest="command", help="Command to execute")

    # create
    create_parser = subparsers.add_parser("create", help="Create a new ADR")
    create_parser.add_argument("--title", required=True, help="ADR title")
    create_parser.add_argument("--background", required=True, help="Background context")
    create_parser.add_argument("--decision", required=True, help="The decision")
    create_parser.add_argument("--consequences", default="", help="Consequences text")
    create_parser.add_argument("--alternatives", default="[]", help="JSON array of alternatives")
    create_parser.add_argument("--scope", default="", help="Comma-separated model element IDs")
    create_parser.add_argument("--status", default="proposed", help="Initial status")

    # list
    list_parser = subparsers.add_parser("list", help="List ADRs")
    list_parser.add_argument("--status", default=None, help="Filter by status")

    # update
    update_parser = subparsers.add_parser("update", help="Update ADR status")
    update_parser.add_argument("--file", required=True, help="ADR filename")
    update_parser.add_argument("--status", required=True, help="New status")
    update_parser.add_argument("--superseded-by", default=None, help="Superseding ADR reference")

    args = parser.parse_args()

    if not args.command:
        parser.print_help()
        sys.exit(1)

    project_root = resolve_project_root()

    if args.command == "create":
        result = create_adr(
            project_root,
            args.title,
            args.background,
            args.decision,
            args.consequences,
            args.alternatives,
            args.scope,
            args.status,
        )
    elif args.command == "list":
        result = list_adrs(project_root, args.status)
    elif args.command == "update":
        result = update_adr(project_root, args.file, args.status, args.superseded_by)
    else:
        result = {"success": False, "error": f"Unknown command: {args.command}"}

    print(json.dumps(result, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()

File: plugins/dev-team/bin/src/lib/archi-decide.ts
/**
 * archi-decide.ts — Architecture Decision Record (ADR) create / list / update.
 * Ported from Python archi-decide.py to MCP-backed TypeScript lib.
 */

import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';

import type {
  AdrStatus,
  ArchiDecideAlternative,
  ArchiDecideCreateInput,
  ArchiDecideCreateResult,
  ArchiDecideListResult,
  ArchiDecideUpdateInput,
  ArchiDecideUpdateResult,
} from '../schemas/archi-decide.schema';

const DECISIONS_DIR = path.join('openspec', 'architecture', 'decisions');
const VALID_STATUSES: readonly AdrStatus[] = ['proposed', 'accepted', 'deprecated', 'superseded'];
const REQUIRED_TEMPLATE_HEADINGS = ['## 背景', '## 决策', '## 后果', '## 备选方案', '## 影响范围'];

function slugify(title: string): string {
  let slug = title.toLowerCase().trim();
  slug = slug.replace(/[^\w\s-]/g, '');
  slug = slug.replace(/[\s_]+/g, '-');
  slug = slug.replace(/-+/g, '-');
  return slug.replace(/^-+|-+$/g, '');
}

function getModuleStartDir(): string {
  if (typeof __dirname === 'string') {
    return __dirname;
  }
  return path.dirname(fileURLToPath(import.meta.url));
}

function resolveAdrTemplatePath(): string | null {
  let dir = getModuleStartDir();
  while (true) {
    const candidate = path.join(dir, 'templates', 'adr.md');
    if (fs.existsSync(candidate)) {
      return candidate;
    }
    const parent = path.dirname(dir);
    if (parent === dir) {
      return null;
    }
    dir = parent;
  }
}

function formatProsCons(value: string | string[] | undefined, label: '优点' | '缺点'): string {
  if (value === undefined) {
    return '';
  }
  if (Array.isArray(value)) {
    return value.map((item) => `- **${label}**: ${item}`).join('\n');
  }
  return `- **${label}**: ${value}`;
}

function renderAlternatives(alternatives: ArchiDecideAlternative[] | undefined): string {
  if (!alternatives?.length) {
    return '\n(无备选方案记录)\n';
  }

  let text = '';
  for (let i = 0; i < alternatives.length; i += 1) {
    const alt = alternatives[i];
    if (!alt) {
      continue;
    }
    text += `\n### 方案 ${i + 1}：${alt.name}\n`;
    text += `- **描述**: ${alt.description ?? alt.name}\n`;
    const pros = formatProsCons(alt.pros, '优点');
    if (pros) {
      text += `${pros}\n`;
    }
    const cons = formatProsCons(alt.cons, '缺点');
    if (cons) {
      text += `${cons}\n`;
    }
  }
  return text;
}

function renderScope(scope: string[] | undefined): string {
  const items = scope?.map((item) => item.trim()).filter(Boolean) ?? [];
  if (items.length === 0) {
    return '- (none)';
  }
  return items.map((item) => `- ${item}`).join('\n');
}

function replaceSectionBody(
  content: string,
  startHeading: string,
  endHeading: string,
  body: string,
): string {
  const startIdx = content.indexOf(startHeading);
  if (startIdx < 0) {
    return content;
  }
  const bodyStart = startIdx + startHeading.length;
  const endIdx = content.indexOf(endHeading, bodyStart);
  if (endIdx < 0) {
    return content;
  }
  const prefix = content.slice(0, bodyStart);
  const suffix = content.slice(endIdx);
  return `${prefix}\n\n${body}\n\n${suffix}`;
}

function replaceTrailingSectionBody(content: string, heading: string, body: string): string {
  const startIdx = content.indexOf(heading);
  if (startIdx < 0) {
    return content;
  }
  const bodyStart = startIdx + heading.length;
  return `${content.slice(0, bodyStart)}\n\n${body}\n`;
}

function renderAdrFromTemplate(
  template: string,
  data: {
    title: string;
    date: string;
    status: AdrStatus;
    background: string;
    decision: string;
    consequences: string;
    alternatives: ArchiDecideAlternative[] | undefined;
    scope: string[] | undefined;
  },
): string {
  let content = template;

  content = content.replace('# ADR: <title>', `# ADR: ${data.title}`);
  content = content.replace('YYYY-MM-DD', data.date);
  content = content.replace(
    '- **状态**: proposed | accepted | deprecated | superseded',
    `- **状态**: ${data.status}`,
  );
  content = content.replace('<描述导致此决策的技术问题或业务需求>', data.background);
  content = content.replace('<宣布并解释所做的决定>', data.decision);
  content = content.replace('- <列出选择带来的好处>', '- (见下方描述)');
  content = content.replace('- <列出选择带来的代价、限制或风险>', '- (见下方描述)');

  if (data.consequences) {
    const marker = '- (见下方描述)';
    const lastIdx = content.lastIndexOf(marker);
    if (lastIdx >= 0) {
      const insertAt = lastIdx + marker.length;
      content = `${content.slice(0, insertAt)}\n\n${data.consequences}${content.slice(insertAt)}`;
    }
  }

  content = replaceSectionBody(
    content,
    '## 备选方案',
    '## 影响范围',
    renderAlternatives(data.alternatives).trim(),
  );

  const scopeBody = `<!-- 引用受此决策影响的模型元素 FQN -->\n\n${renderScope(data.scope)}`;
  content = replaceTrailingSectionBody(content, '## 影响范围', scopeBody);

  return content.endsWith('\n') ? content : `${content}\n`;
}

function isValidStatus(status: string): status is AdrStatus {
  return (VALID_STATUSES as readonly string[]).includes(status);
}

function invalidStatusError(status: string): { success: false; error: string } {
  return {
    success: false,
    error: `Invalid status '${status}'. Must be one of: ${VALID_STATUSES.join(', ')}`,
  };
}

function localDateString(): string {
  const now = new Date();
  const year = now.getFullYear();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  return `${year}-${month}-${day}`;
}

function loadAdrTemplate(): { template: string } | { error: string } {
  const templatePath = resolveAdrTemplatePath();
  if (!templatePath) {
    return { error: 'ADR template not found: templates/adr.md' };
  }

  let template: string;
  try {
    template = fs.readFileSync(templatePath, 'utf-8');
  } catch {
    return { error: `Failed to read ADR template: ${templatePath}` };
  }

  for (const heading of REQUIRED_TEMPLATE_HEADINGS) {
    if (!template.includes(heading)) {
      return { error: `ADR template missing required section: ${heading}` };
    }
  }

  return { template };
}

function writeAdrFile(
  decisionsDir: string,
  filepath: string,
  content: string,
): { success: true } | { success: false; error: string } {
  try {
    fs.mkdirSync(decisionsDir, { recursive: true });
    fs.writeFileSync(filepath, content, 'utf-8');
    return { success: true };
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return { success: false, error: `Failed to write ADR: ${message}` };
  }
}

export function createAdr(
  projectRoot: string,
  input: ArchiDecideCreateInput,
): ArchiDecideCreateResult {
  const statusValue = input.status ?? 'proposed';
  if (!isValidStatus(statusValue)) {
    return invalidStatusError(statusValue);
  }

  const loaded = loadAdrTemplate();
  if ('error' in loaded) {
    return { success: false, error: loaded.error };
  }

  const dateStr = localDateString();
  const filename = `${dateStr}-${slugify(input.title)}.md`;
  const decisionsDir = path.join(projectRoot, DECISIONS_DIR);
  const filepath = path.join(decisionsDir, filename);

  if (fs.existsSync(filepath)) {
    return { success: false, error: `ADR already exists: ${filename}` };
  }

  const content = renderAdrFromTemplate(loaded.template, {
    title: input.title,
    date: dateStr,
    status: statusValue,
    background: input.background,
    decision: input.decision,
    consequences: input.consequences ?? '',
    alternatives: input.alternatives,
    scope: input.scope,
  });

  const written = writeAdrFile(decisionsDir, filepath, content);
  if (!written.success) {
    return written;
  }

  return { success: true, path: filepath, filename };
}

function parseAdrFile(filepath: string, filename: string): ArchiDecideListResult['adrs'][number] {
  const content = fs.readFileSync(filepath, 'utf-8');

  const titleMatch = content.match(/^# ADR:\s*(.+)$/m);
  const title = titleMatch?.[1]?.trim() ?? filename;

  const statusMatch = content.match(/\*\*状态\*\*:\s*(\w+)/);
  const status = statusMatch?.[1]?.trim() ?? 'unknown';

  const dateMatch = content.match(/\*\*日期\*\*:\s*(\S+)/);
  const date = dateMatch?.[1]?.trim() ?? '';

  const scopeSection = content.match(/## 影响范围\s*\n([\s\S]*?)(?=\n##|$)/);
  const scope: string[] = [];
  const scopeBody = scopeSection?.[1];
  if (scopeBody) {
    for (const line of scopeBody.trim().split('\n')) {
      const trimmed = line.trim();
      if (trimmed.startsWith('-')) {
        scope.push(trimmed.replace(/^-\s*/, '').trim());
      }
    }
  }

  return { filename, title, status, date, scope, path: filepath };
}

export function listAdrs(projectRoot: string, statusFilter?: AdrStatus): ArchiDecideListResult {
  const decisionsDir = path.join(projectRoot, DECISIONS_DIR);

  if (!fs.existsSync(decisionsDir) || !fs.statSync(decisionsDir).isDirectory()) {
    return { adrs: [], count: 0 };
  }

  const adrs: ArchiDecideListResult['adrs'] = [];

  let filenames: string[];
  try {
    filenames = fs.readdirSync(decisionsDir).filter((name) => name.endsWith('.md'));
  } catch {
    return { adrs: [], count: 0 };
  }

  for (const filename of filenames) {
    const filepath = path.join(decisionsDir, filename);
    try {
      if (!fs.statSync(filepath).isFile()) {
        continue;
      }
      const entry = parseAdrFile(filepath, filename);
      if (statusFilter && entry.status !== statusFilter) {
        continue;
      }
      adrs.push(entry);
    } catch {
      continue;
    }
  }

  adrs.sort((a, b) => {
    if (a.date !== b.date) {
      return b.date.localeCompare(a.date);
    }
    return b.filename.localeCompare(a.filename);
  });

  return { adrs, count: adrs.length };
}

function applyStatusUpdate(
  content: string,
  status: AdrStatus,
  supersededBy: string | undefined,
): { content: string; oldStatus: string } | { error: string } {
  const oldStatusMatch = content.match(/\*\*状态\*\*:\s*(\w+)/);
  const oldStatus = oldStatusMatch?.[1]?.trim() ?? 'unknown';

  const statusLineMatch = content.match(/(\*\*状态\*\*:\s*)\w+/);
  if (!statusLineMatch || statusLineMatch.index === undefined) {
    return { error: 'ADR file missing **状态** field' };
  }

  let updated =
    content.slice(0, statusLineMatch.index) +
    `**状态**: ${status}` +
    content.slice(statusLineMatch.index + statusLineMatch[0].length);

  if (status === 'superseded' && supersededBy && !updated.includes('**取代者**')) {
    const statusEnd = updated.indexOf('\n', updated.indexOf(`**状态**: ${status}`));
    const refLine = `\n- **取代者**: ${supersededBy}`;
    if (statusEnd > 0) {
      updated = updated.slice(0, statusEnd) + refLine + updated.slice(statusEnd);
    }
  }

  return { content: updated, oldStatus };
}

function resolveAdrFilePath(
  projectRoot: string,
  file: string,
): { filepath: string } | { error: string } {
  const decisionsDir = path.resolve(projectRoot, DECISIONS_DIR);
  const filepath = path.resolve(decisionsDir, file);
  const relative = path.relative(decisionsDir, filepath);
  if (relative.startsWith('..') || path.isAbsolute(relative)) {
    return { error: `Path '${file}' is outside decisions directory` };
  }
  return { filepath };
}

export function updateAdr(
  projectRoot: string,
  input: ArchiDecideUpdateInput,
): ArchiDecideUpdateResult {
  const statusValue = input.status;
  if (!isValidStatus(statusValue)) {
    return invalidStatusError(statusValue);
  }

  if (statusValue === 'superseded' && !input.superseded_by) {
    return {
      success: false,
      error: "Status 'superseded' requires superseded_by reference",
    };
  }

  const resolved = resolveAdrFilePath(projectRoot, input.file);
  if ('error' in resolved) {
    return { success: false, error: resolved.error };
  }
  const filepath = resolved.filepath;
  if (!fs.existsSync(filepath) || !fs.statSync(filepath).isFile()) {
    return { success: false, error: `ADR not found: ${input.file}` };
  }

  let content: string;
  try {
    content = fs.readFileSync(filepath, 'utf-8');
  } catch (err) {
    const message = err instanceof Error ? err.message : String(err);
    return { success: false, error: `Failed to read ADR: ${message}` };
  }

  const applied = applyStatusUpdate(content, statusValue, input.superseded_by);
  if ('error' in applied) {
    return { success: false, error: applied.error };
  }

  const written = writeAdrFile(path.dirname(filepath), filepath, applied.content);
  if (!written.success) {
    return written;
  }

  return {
    success: true,
    path: filepath,
    old_status: applied.oldStatus,
    new_status: statusValue,
  };
}
